### PR TITLE
[kimi k3] Support sample packing in KDA

### PR DESCRIPTION
**Pull request by JavaZeroo**
## What changes

- `KimiK3Model.get_attention_masks` returns a mask dict, like Qwen3.5. MLA layers read "quadratic_attention" and KDA layers read "kda". The "kda" entry is the VarlenMetadata that Attention Gym's causal_conv1d and chunk_kda use to reset the conv window and the recurrent state at each document start. Each block picks its own entry by `attn_mask_key`. The signature is the same as `Decoder.get_attention_masks` after #4156.
- The multimodal packer numbers its padding positions from the first padded token, modulo max_context_length.
- `kimi_k3_debugmodel` takes `seq_len` like every other debug config and sizes its rows from the model spec, so it runs 2048-token rows instead of a fixed 256.

The text path and the helper itself landed in #4156. This PR only touches the multimodal packer, the multimodal collator and Kimi K3.

## Verification

**Packing on the debug config.** `kimi_k3_debugmodel` rows are 2048 tokens and one cc12m sample is about 210 tokens, so without the packer a row holds one document and the rest is padding. With `MMSamplePackingConfig` wrapped around the dataset:

| | Documents per row | Padding tokens per row |
|---|---|---|
| No packer | 1 | 1837 |
| Packer | 16 | 3 |

**End to end on B200.** `kimi_k3_debugmodel`, 10 steps, `--debug.seed 42 --debug.deterministic`. With the packer, loss went from 12.50 to 3.26 over 10 steps and grad_norm from 15.1 to 2.2. Without the packer, loss went from 12.45 to 4.50. These numbers come from the previous revision of this branch; the K3 wiring is unchanged. TODO: re-run on this revision.

Kimi K3 does not match main bitwise on rows with padding, and should not: on main the KDA layers receive no offsets at all.

## TODO

- [ ] `get_attention_masks` in Kimi K3 and Qwen3.5 are near-identical and should be shared; needs a design.

### CODE REVIEW

File: torchtitan/models/common/attention.py [hunk 1]
```diff
@@ -655,6 +658,49 @@ def create_varlen_metadata_for_document(
     )
 
 
+def create_varlen_metadata_for_packed_document(
```
Comment (tianyu-l) [inline torchtitan/models/common/attention.py]: We always do packing, without exception. We should always assume padding. And only `create_varlen_metadata_for_document` should be used.

It sounds this treatment is due to the fact that dataloader always specify position as 0 on padded tokens so that the metadata is misleading. Without looking into details, I think we probably should let dataloader pass padding mask back to trainer so that there's no need to implicitly infer. Specifically
- remove this new function and the special treatment in qwen3.5
- change dataloader behavior to also produce padding mask (see related discussions in https://github.com/pytorch/torchtitan/pull/4168#issuecomment-5348820598)
- fix existing `create_varlen_metadata_for_document`

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/models/common/attention.py (hunk 1 above)
Comment (JavaZeroo) [reply torchtitan/models/common/attention.py]: thanks for reviewing, done:
- removed `create_varlen_metadata_for_packed_document` and the special treatment in qwen3.5
- the packers now derive a padding_mask from Grain's input_ids_segment_ids and the
  collators extend it with the tail they append
- `create_varlen_metadata_for_document` now takes that mask. 

And I found the offsets were also being rebuilt per layer. FLA memoizes its varlen index tables on argument identity, so the first conv of every layer missed that cache and paid a rebuild plus a d2h. (I originally fixed it for both K3 and qwen3.5, but after #4351 K3 no longer uses FLA, so it only applies to qwen3.5 now.)

**Comment by JavaZeroo** [comment]
@tianyu-l @shuhuayu All the review comments are addressed, and I verified the KDA varlen path end-to-end on a B200. Ready for another look whenever you have time.

File: torchtitan/models/qwen3_5/model.py [hunk 2]
```diff
@@ -445,43 +449,23 @@ def preprocess_inputs(
...
-            ]
-        )
-        first_token = torch.arange(positions.shape[0], device=positions.device) == 0
-        sequence_starts = ((positions == 0) & followed_by_one) | first_token
-        sequence_positions = torch.where(sequence_starts, 0, 1)
         deltanet_metadata = create_varlen_metadata_for_document(
-            sequence_positions,
+            positions,
+            padding_mask=padding_mask,
             include_host_offsets=True,
```
Comment (tianyu-l) [inline torchtitan/models/qwen3_5/model.py:459]: this field is removed, please rebase

File: torchtitan/models/qwen3_5/model.py [hunk 3]
```diff
@@ -445,43 +449,23 @@ def preprocess_inputs(
     def get_attention_masks(
         self,
         positions: torch.Tensor,
+        padding_mask: torch.Tensor | None = None,
```
Comment (tianyu-l) [inline torchtitan/models/qwen3_5/model.py:452]: please make it a kwarg, instead of positional (we probably should do the same for `positions` but the scope would be too much, can do it later)

File: torchtitan/models/qwen3_5/gdn.py [hunk 4]
```diff
@@ -363,25 +353,16 @@ def forward(
         *,
         key_head_dim: int,
         value_head_dim: int,
-        cu_seqlens_host: tuple[int, ...] | None = None,
+        cu_seqlens_cpu: torch.Tensor | None = None,
```
Comment (tianyu-l) [inline torchtitan/models/qwen3_5/gdn.py:356]: similar

File: torchtitan/models/kimi_k2_7/model.py
Comment (tianyu-l) [inline torchtitan/models/kimi_k2_7/model.py:125]: why not feeding `padding_mask` here?

File: torchtitan/models/common/attention.py [hunk 5]
```diff
@@ -593,16 +594,19 @@ def create_attention_mask(*args, **kwargs):
 def create_varlen_metadata_for_document(
```
Comment (tianyu-l) [inline torchtitan/models/common/attention.py:594]: could you take a look at https://github.com/pytorch/torchtitan/pull/4156 and see what's the difference?

Comment (JavaZeroo) [reply torchtitan/models/common/attention.py:594]: Looked at #4156. It already implements the text-path padding_mask, and it keeps every segment within max_context_length, which mine doesn't. I think I'll wait for it to merge and then rebase this PR on top of it, keeping only the multimodal packer/collator and the  Kimi K3 mask wiring.

**Comment by shuhuayu** [comment]
Hi @JavaZeroo, https://github.com/pytorch/torchtitan/pull/4156 has merged, can you rebase?

**Comment by JavaZeroo** [comment]
> Hi @JavaZeroo, https://github.com/pytorch/torchtitan/pull/4156 has merged, can you rebase? 

Sure, im currently working on it.

File: torchtitan/models/kimi_k3/model.py [hunk 6]
```diff
@@ -43,6 +45,10 @@
 from .moe import KimiFeedForward, KimiLatentMoE
 from .vision_encoder import KimiK3VisionEncoder
 
+# MLA layers read "quadratic_attention" (a BlockMask under flex, shared offsets
+# under varlen); KDA layers read "kda" (document offsets).
```
Comment (shuhuayu) [inline torchtitan/models/kimi_k3/model.py]: nit: remove.

File: torchtitan/hf_datasets/multimodal/mm_collator.py [hunk 7]
```diff
@@ -81,11 +81,23 @@ def collate_images(
...
         self,
         batch: list[dict[str, Any]],
-    ) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    ) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
         """Concatenate whole samples and pad only the token-batch tail."""
         input_ids = torch.cat([sample["input_ids"] for sample in batch])
         labels = torch.cat([sample["labels"] for sample in batch])
         positions = torch.cat([sample["positions"] for sample in batch])
+        # Sample packing pads inside each row, so carry the packer's own mask
+        # through and extend it with this batch's tail padding.
```
Comment (shuhuayu) [inline torchtitan/hf_datasets/multimodal/mm_collator.py]: nit: remove.

**Review by shuhuayu — APPROVED** [review]
lgtm, some nit comments.

**Review by tianyu-l — APPROVED** [review]
